### DISCUSSION TITLE
Study group tags

### DISCUSSION THREAD

**Discussion by debritto** [ezra-bible-app/ezra-bible-app#540] [Ideas]
Using Ezra Bible App, I realize that sometimes is important to create tags for a specific kind of Biblical study. For example, suppose that I want to create tags related to a study about the "sovereingty of God", for this I could create tags like: "God's Power", "Miracles", "Prophecy" etc. 
Now suppose that I want to conduct a parallel study about the use of money in the Bible, for this I could create tags like: "Offerings", "Widows", "Missions" etc.

Now, if I want to explore my findings it will be easier if there is an option to select tags grouped by study. So instead of this:

![Tags per study](https://user-images.githubusercontent.com/9845780/142673987-9d371419-02d6-4c56-aef2-ecba1c9eb154.png)

I'll have something like this:

- Study about the Sovereingty of God
[]  God's power
[] Miracles
[] Prophecy

- Study about mooney in the Bible
[] Offerings
[] Widows
[] Missions

OBS: The "study types" or group of tags is not a tag itself, it is just a way to group them and doesn't need to appear as individual tags.

Warm regards.

**Comment by tobias-klein** [comment]
This could be done in two ways from a data modeling perspective.

1. Utilize arbitrary categories (see #235). The categories could also be broader topics themselves.
2. Introduce something that I used to call *meta tags*. I actually had this once in the web-based version of Ezra, but got rid of it when porting to Electron. The difference between a *meta tag* and a *category* is that in case of a meta tag you could assign multiple meta tags to the same tag, so it would be an n:m relationship between meta tags and tags, but in case of categories it would be 1:m.

Depending on what data model you use for this the UI would also end up being different. The UI that I used to have for the meta tags was showing them as a separate list above the tag list. There you would be able to open a meta tag like a folder and then once you did that you would see additional buttons in the tag list that would allow you to add a tag to the respective "meta tag" folder.

On top of that this could then also be considered for the selection of tagged verse lists.

**Comment by debritto** [comment]
Hi @tobias-klein !

I think that the idea of meta tags is the best one, because is it possible to assigns the same tag to different meta tags (study groups). It's common to reuse some tags in different studies. 

The Meta Tags concept could enable deep textual study techniques. Just for curiosity's sake a good coding (tagging) systems lies at the heart of qualitative analysis, that mostly deal with texts. I think that Ezra Bible App could be used as a serious biblical qualitative analysis resource. Please, take a look at this software -> [MaxQDA](https://www.maxqda.com/what-is-maxqda)

It's one of the best for qualitative analysis, but it's not good for bible studies because the necessity to use different translations and strong numbers. But look at their coding system window:

![image](https://user-images.githubusercontent.com/9845780/142914052-e2eb1a31-9196-4079-b5ee-32a1b730994f.png)

Is it possible to use even different tag colors.
With love of Christ,
Christian

**Comment by tobias-klein** [comment]
Yeah, so meta tags do express what you've been thinking :) Good to hear that.

One challenge with this is to keep it as simple and as easy as possible from a UX perspective. The kind of study you describe (qualitative analysis) is certainly rather something for the *researcher* than for the *average casual user*. The *average casual user* may not even need meta tags.

With that said, we should probably find a way to offer this feature to the *researcher* while not bloating the UI to much for the *average casual user* who may not even need that feature.

@zhuiks Any thoughts on this?

**Comment by debritto** [comment]
I agree with you @tobias-klein . 

One idea is to keep the UI as it is today, so users can create tags as simple as possible. Then one could create a tag and rather than assign it to verses one could "drag and drop" another tag to it, automatically turning it into a meta-tag.
If the user try to drag and drop a tag with verses attached to another the App could inform that the operation is only possible with "empty" tags. 

What do you think?

**Comment by tobias-klein** [comment]
That sounds a bit too implicit (not so obvious to the user), honestly. I'll try to dig out my old ruby on rails app one of these days and see how it looks there. I once already implemented this and I can show how it looked back then. At least as a starting point.

**Comment by debritto** [comment]
The other possibility is to put an option "Enable Advanced Features" here:

![image](https://user-images.githubusercontent.com/9845780/142943653-e939504c-6d7d-445c-885d-0b6da61ea448.png)

This option could enable "Export all to CSV", "Show Verse Occurrences Per Tag" and, until now, "Create Meta Tags" features.
This could maintain a simpler or a more advanced UI.

**Comment by tobias-klein** [comment]
I just pulled up my old web-based Ezra version from a virtual machine I still have. This is how far I got with meta tags some years back ... it was essentially a basic "folder-like" UI that allowed to assign/unassign tags to a meta tag.

![old_ezra_meta_tags](https://user-images.githubusercontent.com/17739895/143106191-b0f5f528-b71f-40d3-8779-1c7c786c92de.gif)

**Reply by debritto** [reply]
This is it!
Great idea. Works great in this way.

**Comment by tobias-klein** [comment]
Going to convert this into a discussion until we have settled on a clear implementation concept.

**Comment by zhuiks** [comment]
Hi guys,
Here are my thoughts:
1. The concept is great and worth of our implementation efforts.
2. I wouldn't add any extra toolbox/windows for meta-tags/categories. Although, it can be a subsection of current tag window.
3. I wouldn't add any "advance feature" switch in options. I think it brings extra complexity on implementation level as well as confuses end users how they can decide if they need advance features (Personally, I like to see all the options available, so I always turn advanced features on, although I might be new to the app).
4. Tree view seems to be the natural implementation of those type of things. If we go with meta-tags, we have to decide how it can be implemented with the tree view. @debritto Do you have any examples from other interfaces?

There is similar idea with grouping in graphic editors. Here is how it looks in Figma. One can create groups and groups of groups from selected elements (blue background).
![Screenshot from 2021-11-25 08-17-18](https://user-images.githubusercontent.com/3584119/143460950-343e4933-c95a-440a-9e23-459eb6fc01b6.png)

**Comment by debritto** [comment]
Hi @zhuiks !

Yes. I think you're right, if we create an advanced menu it could be confusing and bring an unnecessary complexity to future implementations. You have good points!

Here are some examples of tag trees from another apps. 

This one is from a german open source project called CATMA:
![image](https://user-images.githubusercontent.com/9845780/143486255-d11bf133-30c4-49a2-9a84-5c9613f41123.png)
This one is from a german open source project called CATMA. Here it's possible to add Tagsets, Tags and Subtags.

Here we have F4 Analyse:

![image](https://user-images.githubusercontent.com/9845780/143486458-235cc4dd-ae06-4e7c-b514-491731f9983a.png)
This is another German software (Germans are very good on software development ;) ), is it possible to put colors on tags. Here are a small overview of this [software](https://www.audiotranskription.de/en/f4analyse/).

Here is a solution called QDAMiner:
![image](https://user-images.githubusercontent.com/9845780/143486717-ac897b36-673c-41c5-985b-bec85ba59696.png)
It use a tree like input.

Warm regards!

**Reply by tobias-klein** [reply]
Good discussion! Thanks for the examples @debritto 
Thanks for the feedback @zhuiks.

I agree that we should keep it one box/panel if possible.
As an alternative to a tree view, I can also imagine a folder like navigation. I imagine this to be easier from a navigational perspective especially when having lots of tags under one parent. This would also be easier to implement.

Here are some additional clarifications:
1) **Data model**. Should it be possible that one tag can have multiple meta tags / tag groups?
2) **Data model**. In my mind we should have a clear separation of concerns between *tags* and *tag groups*. Which means that you do not assign *tag groups* to verses, but only *tags*. And the sole purpose of *tag groups* is to organize *tags* with more overview and allow for filtering/selecting based on higher-level grouping.
3) **Naming**. What English name would you prefer? *Meta tags* make sense to me, but may be too technical for typical users. How about *tag groups*?
4) **Number of hierarchy levels**. Should we constrain that? To how many levels? I believe two tag group levels would be the maximum, but we could even limit it to one, to make it simpler.

**Reply by debritto** [reply]
If a folder like navigation is easier to use, we can go for it. I think works well.

1. **Data model.** Yes. Its important to allow a tag to be a part of more than one groups - especially because we can reuse some tags in different studies. The other practical example, will be the possibility to have tags like "Textual type" -> (Epistles; Narrative; Poetry; Phrophecy)
2. **Data model.** Exactly!
3. **Naming.** "Tag group" is very good.
4. I think that only one level is sufficient. I mean, a tag group with its tags.

**Reply by tobias-klein** [reply]
@debritto I have been starting to work on "tag groups" and I am planning to publish this new feature in the next release 1.6. You can follow the progress in PR #641.

**Reply by debritto** [reply]
Thank you so much @tobias-klein!
I've been working with Ezra Bible App everyday and it's a awesome app. As my Tg list increase and different Bible studies are being made the Tag Group feature is more than welcome. Waarm regards from Brazil.
